src/app/hooks/reactQuery.ts

```diff
@@ -47,18 +67,58 @@ export function usePrograms() {
  * @returns React Query useQuery Hook for all majors and minors
  */
 export function useUnqiuePrograms() {
+  const queryClient = useQueryClient();
+
   return useQuery({
-    queryKey: ["allUniquePrograms"],
+    queryKey: ["uniquePrograms"],
     queryFn: async () => {
       const res = await getPrograms();
       return filterRedundantPrograms(res);
     },
+    // Use cache to avoid refetching data
+    initialData: () => {
+      const programs: Program[] | undefined = queryClient.getQueryData([
+        "programs",
+      ]);
+      return programs ? filterRedundantPrograms(programs) : [];
+    },
     placeholderData: [],
+    staleTime: Infinity, // Programs are static
+    refetchOnMount: false,
+  });
+}
+/**
+ * A React Query hook to fetch all unique majors and minors for a specific
+ * catalog year
+ * @param catalogYear Year of selected catalog
+ * @returns React Query useQuery Hook for all majors and minors
+ */
+export function useUnqiueProgramsInYear(catalogYear: string) {
+  const queryClient = useQueryClient();
+
+  return useQuery({
+    queryKey: ["uniquePrograms", catalogYear],
+    queryFn: async () => {
+      const res = await getPrograms(catalogYear);
+      return filterRedundantPrograms(res);
```

Comment:
Any reason we can't move the `filterRedundantPrograms` to the server?

Reply:
We could definitely do that, would probably have to write a raw SQL query as prisma doesn't support the DISTINCT keyword in SQL. This way is nice as it saves database calls and derives the data from existing cache data if it exists, but does require the client to perform the work.

Up to you, if you think a dedicated server action with a SQL DISTINCT query is good going forward I can add it to this PR.

Reply:
If there is no need for duplicate programs returned by the `getPrograms` server action, we can leverage the distinct field in the parameter object of `findMany` and avoid having to use `filterRedundantPrograms` altogether as follows:
```ts
const distinctCities = await prisma.user.findMany({
  select: {
    city: true,
    country: true,
  },
  distinct: ['city'],
}) 
```

Reply:
Prisma does have a distinct option but it doesn't actually use SQL Distinct but rather in-memory sorting after the query which is not as well optimized ([prisma docs explains it here](https://www.prisma.io/docs/orm/prisma-client/queries/aggregation-grouping-summarizing#distinct-under-the-hood)). Not a super big performance hit so I guess it doesn't matter to much but may be beneficial to avoid.

Also there is a use case for both all programs and distinct programs. In my `ProgramSelector.tsx` I reduced all the db calls to just one action and derived the necessary data so unique would not work here. I'll add a server action to my PR for the permissions modal that uses distinct programs, but maybe its a better tradeoff to move redundancy filtering to the client and utilize the react query cache, not sure.

``` typescript
/**
 * A React Query hook to fetch all unique majors and minors
 * @returns React Query useQuery Hook for all majors and minors
 */
export function useUniquePrograms() {
  const queryClient = useQueryClient();

  return useQuery({
    queryKey: ["uniquePrograms"],
    queryFn: async () => {
      const res = await getPrograms();
      return filterRedundantPrograms(res);
    },
    // Use cache to avoid refetching data
    initialData: () => {
      const programs: Program[] | undefined = queryClient.getQueryData([
        "programs",
      ]);
      return programs ? filterRedundantPrograms(programs) : [];
    },
    placeholderData: [],
    staleTime: Infinity, // Programs are static
    refetchOnMount: false,
  });
}
```